EditorState fonts should only sync with the UI Process when an editor requests the data

gavin-apple opened the pull request:
#### 441f90285c076d335b609f91d1ecfdabf69cb859
<pre>
EditorState fonts should only sync with the UI Process when an editor requests the data
<a href="https://bugs.webkit.org/show_bug.cgi?id=299027">https://bugs.webkit.org/show_bug.cgi?id=299027</a>
<a href="https://rdar.apple.com/160600679">rdar://160600679</a>

Reviewed by Abrar Rahman Protyasha.

Currently, we eagerly sync EditorState fonts with the UI process. However,
this data is rarely needed. This change moves to a model where we only request
the font data lazily.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/FontAttributes.mm
* Source/WebKit/Shared/EditorState.h:
* Source/WebKit/Shared/EditorState.serialization.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updateFontAttributesAfterEditorStateChange):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::updateFontManagerIfNeeded):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::editorState const):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/FontAttributes.mm:
(-[FontTextStyleUIDelegate _webView:didChangeFontAttributes:]):
(-[FontAttributesListener _webView:didChangeFontAttributes:]):
(-[TestWKWebView fontAttributesAfterNextPresentationUpdate]):
(TestWebKitAPI::TEST(FontAttributes, FontAttributesAfterChangingSelection)):

Canonical link: <a href="https://commits.webkit.org/300451@main">https://commits.webkit.org/300451@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f7d95f9a7f02d76b4a7f3add7e8d22da413a014c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122463 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42171 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/32856 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129061 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/74562 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/3eeb9756-465e-4f16-b149-ea184babd400) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124339 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42889 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50764 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93087 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61835 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/45242bd9-6ef2-4831-ae4d-af51f4b6479e) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125415 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34207 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109653 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73738 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33195 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27809 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72548 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/103848 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28020 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/131791 "Built successfully") | | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/120989 "Build is in progress. Recent messages:") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49404 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37600 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101629 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/49778 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105873 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101498 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25791 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46876 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25014 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/46172 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49262 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55010 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/48729 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52079 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50411 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3f766bb1d1dd7816ef95e04c5a34fdc4806900a3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50900

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/151434c96f9b775172cf9b7e4d6d2a38a33add2c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50900

gavin-apple commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3625,15 +3622,9 @@ void WebPageProxy::requestFontAttributesAtSelectionStart(CompletionHandler<void(
-        internals().cachedFontAttributesAtSelectionStart = attributes;
-        callback(attributes);
-    });
+    auto sendResult = protectedLegacyMainFrameProcess()->sendSync(Messages::WebPage::RequestFontAttributesAtSelectionStart(), webPageIDInMainFrameProcess());
```
I think we should revert this to async and accept the small amount of latency. Currently just checking that this addresses all issues with tests on EWS.

whsieh replied:
Indeed — we shouldn't be adding any new sync IPC messages (in particular, sync messages from the UI process to the web content process).

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b142919474c2803ecb314b560095893acb9712d7. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50900

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d2f0b9904603d8f1847aed3910a7e51eb8ab6ce5. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50900

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/FontAttributes.mm:
```diff
@@ -105,7 +109,11 @@ - (void)selectElementWithIdentifier:(NSString *)identifier
 
 - (NSDictionary *)fontAttributesAfterNextPresentationUpdate
 {
+    ((FontAttributesListener *)self.UIDelegate)->_attributeUpdated = NO;
```
```suggestion
    checked_objc_cast<FontAttributesListener>(self.UIDelegate)->_attributeUpdated = NO;
```

gavin-apple committed:
EditorState fonts should only sync with the UI Process when an editor requests the data
https://bugs.webkit.org/show_bug.cgi?id=299027
rdar://160600679

Reviewed by Abrar Rahman Protyasha.

Currently, we eagerly sync EditorState fonts with the UI process. However,
this data is rarely needed. This change moves to a model where we only request
the font data lazily.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/FontAttributes.mm
* Source/WebKit/Shared/EditorState.h:
* Source/WebKit/Shared/EditorState.serialization.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updateFontAttributesAfterEditorStateChange):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::updateFontManagerIfNeeded):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::editorState const):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/FontAttributes.mm:
(-[FontTextStyleUIDelegate _webView:didChangeFontAttributes:]):
(-[FontAttributesListener _webView:didChangeFontAttributes:]):
(-[TestWKWebView fontAttributesAfterNextPresentationUpdate]):
(TestWebKitAPI::TEST(FontAttributes, FontAttributesAfterChangingSelection)):

Canonical link: https://commits.webkit.org/300451@main

webkit-commit-queue commented:
Committed 300451@main (441f90285c07): <https://commits.webkit.org/300451@main>

Reviewed commits have been landed. Closing PR #50900 and removing active labels.
